mlflow/pipelines/steps/train.py

```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
-                    estimator,
-                    f"{self.name}/estimator",
-                    signature=estimator_schema,
-                    code_paths=code_paths,
+                # wrap training in objective fn
+                def objective(hyperparameter_args):
+                    # log as a child run
+                    with mlflow.start_run(tags=tags, nested=True):
+                        # create unfitted estimator from yaml
+                        estimator = estimator_fn(
+                            dict(estimator_hardcoded_params, **hyperparameter_args)
+                        )
+                        sample_fraction = (
+                            float(tuning_params["sample_fraction"])
+                            if "sample_fraction" in tuning_params
+                            else 1.0
+                        )
+                        X_train_sampled = X_train.sample(frac=sample_fraction, random_state=42)
+                        y_train_sampled = y_train.sample(frac=sample_fraction, random_state=42)
+                        # fit estimator to training
+                        estimator.fit(X_train_sampled, y_train_sampled)
+                        if hasattr(estimator, "best_score_P"):
+                            mlflow.log_metric("best_cv_score", estimator.best_score_)
+                        if hasattr(estimator, "best_params_"):
+                            mlflow.log_params(estimator.best_params_)
+
+                        code_paths = [os.path.join(self.pipeline_root, "steps")]
+                        estimator_schema = infer_signature(
+                            X_train_sampled, estimator.predict(X_train_sampled.copy())
+                        )
+                        logged_estimator = mlflow.sklearn.log_model(
+                            estimator,
+                            f"{self.name}/estimator",
+                            signature=estimator_schema,
+                            code_paths=code_paths,
+                        )
+                        # evaluate on primary metric
+                        eval_result = mlflow.evaluate(
+                            model=logged_estimator.model_uri,
+                            data=validation_df,
+                            targets=self.target_col,
+                            model_type="regressor",
+                            evaluators="default",
+                            dataset_name="validation",
+                            custom_metrics=_load_custom_metric_functions(
+                                self.pipeline_root,
+                                self.evaluation_metrics.values(),
+                            ),
+                            evaluator_config={
+                                "log_model_explainability": False,
+                            },
+                        )
+
+                        # return +/- metric
+                        return eval_result.metrics[self.primary_metric]
```

Comment:
Hyperopt needs to know whether the objective is to minimize or maximize this metric based on the "greater_is_better" value associated with the metric. Can we incorporate that into this PR? This needs to be supported for predefined metrics and custom metrics.

Reply:
@dbczumar i'm not quite sure how to solve this. the only documentation i see about `greater_is_better` is in [sklearn.metrics.make_scorer](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html), which i don't think is useful here.

what i'm looking for is something like `is_metric_greater_is_better(metric_name)`, as in [catboost](https://github.com/catboost/catboost/issues/1890), but i don't think this is supported in sklearn.

Reply:
Ah, the user specifies `greater_is_better` for custom metrics here: https://github.com/mlflow/mlp-regression-template/blob/74284ee13d907b0b427fd76d144986ac333d520f/pipeline.yaml#L85. We just need to read this value.

For builtin metrics, `greater_is_better` is defined here: https://github.com/mlflow/mlflow/blob/7219771141c7140fd68f6ed3a5b649e5ad1dab13/mlflow/pipelines/utils/metrics.py#L40-L44